### PR TITLE
Background web server does not honour port parameter

### PR DESCRIPTION
**Pull request by hummypkg**
The optional port parameter to =h& is ignored.
This is one way to fix it.

Before:

    [0x00413aa0]> =h& 8888
    Background http server started.
    [0x00413aa0]> Starting http server...
    open http://127.0.0.1:9090/
    r2 -C http://127.0.0.1:9090/cmd/

After:

    [0x004004d0]> =h& 8888
    Background http server started.
    [0x004004d0]> Starting http server...
    open http://127.0.0.1:8888/
    r2 -C http://127.0.0.1:8888/cmd/

### CODE REVIEW

**Comment by Maijin** [comment]
This is configurable with `e http.port`

**Comment by hummypkg** [comment]
Yes, but the help says that the port can also be specified on the command line and it does not work without this patch.

    http server:
    | =h port            listen for http connections (r2 -qc=H /bin/ls)
    | =h-                stop background webserver
    | =h*                restart current webserver
    | =h& port           start http server in background)

File: libr/core/rtr.c [hunk 1]
```diff
@@ -810,7 +814,10 @@ R_API int r_core_rtr_http(RCore *core, int launch, const char *path) {
 			eprintf ("TODO: Use different eval environ for scr. for the web\n");
 			eprintf ("TODO: Visual mode should be enabled on local\n");
 		} else {
-			HttpThread ht = { core, launch, path };
+			char *tpath = path + 1;
+			while (*tpath && isspace(*tpath))
+				tpath++;
+			HttpThread ht = { core, launch, strdup(tpath) };
```
Comment (radare) [inline libr/core/rtr.c]: missing space before (

File: libr/core/rtr.c [hunk 2]
```diff
@@ -810,7 +814,10 @@ R_API int r_core_rtr_http(RCore *core, int launch, const char *path) {
 			eprintf ("TODO: Use different eval environ for scr. for the web\n");
 			eprintf ("TODO: Visual mode should be enabled on local\n");
 		} else {
-			HttpThread ht = { core, launch, path };
+			char *tpath = path + 1;
+			while (*tpath && isspace(*tpath))
```
Comment (radare) [inline libr/core/rtr.c]: use r_str function to skip spaces

File: libr/core/rtr.c [hunk 3]
```diff
@@ -810,7 +814,10 @@ R_API int r_core_rtr_http(RCore *core, int launch, const char *path) {
 			eprintf ("TODO: Use different eval environ for scr. for the web\n");
 			eprintf ("TODO: Visual mode should be enabled on local\n");
 		} else {
-			HttpThread ht = { core, launch, path };
+			char *tpath = path + 1;
```
Comment (radare) [inline libr/core/rtr.c]: char *tpath = r_str_chop (path+1);

File: libr/core/rtr.c (hunk 1 above)
Comment (radare) [inline libr/core/rtr.c]: isnt this leaking?

File: libr/core/rtr.c [hunk 4]
```diff
@@ -783,7 +783,7 @@ static int r_core_rtr_http_thread (RThread *th) {
 	HttpThread *ht = th->user;
 	if (!ht || !ht->core) return false;
 	ret = r_core_rtr_http_run (ht->core, ht->launch, ht->path);
-	free (ht->path);
+	free ((void *)ht->path);
```
Comment (radare) [inline libr/core/rtr.c]: that cast sounds scary

File: libr/core/rtr.c [hunk 5]
```diff
@@ -814,7 +814,7 @@ R_API int r_core_rtr_http(RCore *core, int launch, const char *path) {
 			eprintf ("TODO: Use different eval environ for scr. for the web\n");
 			eprintf ("TODO: Visual mode should be enabled on local\n");
 		} else {
-			char *tpath = path + 1;
+			char *tpath = (char *) path + 1;
```
Comment (radare) [inline libr/core/rtr.c]: this is even more scary

File: libr/core/rtr.c (hunk 4 above)
Comment (hummypkg) [inline libr/core/rtr.c]: free() doesn't like const char * arguments. I can cast to char* if you prefer.
The original code didn't strdup(path) when it passed it to the thread which looks like a race condition to me.

File: libr/core/rtr.c (hunk 3 above)
Comment (hummypkg) [inline libr/core/rtr.c]: Thanks - I'll use r_str_trim_const() since these are const char and I'm only looking to trim leading whitespace.

**Comment by hummypkg** [comment]
Here's an alternative patch but doesn't this have a race condition with the path argument in the thread?

    --- a/libr/core/rtr.c
    +++ b/libr/core/rtr.c
    @@ -376,7 +376,7 @@ static int r_core_rtr_http_run (RCore *core, int launch, const char *path) {
                    return false;
            }
    
    -       if (path && atoi (path)) {
    +       if (path && (atoi (path) || (*path == '&' && atoi (path + 1))) {
                    port = path;
                    path = NULL;
            } else {

File: libr/core/rtr.c [hunk 6]
```diff
@@ -783,7 +783,7 @@ static int r_core_rtr_http_thread (RThread *th) {
 	HttpThread *ht = th->user;
 	if (!ht || !ht->core) return false;
 	ret = r_core_rtr_http_run (ht->core, ht->launch, ht->path);
-	free ((void *)ht->path);
+	free ((char *)ht->path);
```
Comment (radare) [inline libr/core/rtr.c]: the path field in the HttpThread struct must be char * instead of const char *. and then you can remove the casts

**Comment by radare** [comment]
why a rac condition here? the whole r2 is not thread-safe, so using the threading functionalities will probably cause segfaults if using the shell and the web at the same time.  i dont wanna mess the code of r2 with mutexes everywhere, so i will try to find a less intrusive solution

**Comment by hummypkg** [comment]
The thread receives a pointer to path.
As far as I can tell, r_core_cmd_subst could free that memory before the thread uses it.

**Comment by radare** [comment]
Thanks! merged

**Pull request closed without merging**